specifyweb/businessrules/views.py

```diff
@@ -1 +1,288 @@
+                "application/json": {
+                    "schema": {
+                        "type": "object",
+                        "properties": {
+                            "rules": {
+                                "type": "array",
+                                "description": "The array of uniqueness rules for a given table",
+                                "items": UniquenessRuleSchema
+                            }
+                        }
+                    }
+                }
+            }
+        },
+        "responses": {
+            "201": {
+                "description": "Uniqueness rules properly updated and/or created",
+                "content": {
+                    "application/json": {
+                        "schema": {
+                            "type": "object"
+                        }
+                    }
+                }
+            }
+        }
+    }
+})
+@login_maybe_required
+@require_http_methods(['GET', 'PUT', 'POST'])
+@transaction.atomic
+def uniqueness_rule(request, discipline_id):
+    data = {}
+
+    try:
+        model = request.GET["model"]
+    except:
+        model = None
+
+    if request.method == 'GET':
+        rules = UniquenessRule.objects.filter(discipline=discipline_id)
+        for rule in rules:
+            rule_fields = rule.splocalecontaineritems.filter(
+                uniquenessrule_splocalecontaineritem__isScope=0)
+            scope = rule.splocalecontaineritems.filter(
+                uniquenessrule_splocalecontaineritem__isScope=1)
+
+            table = rule_fields[0].container.name
+            if model is not None and table.lower() != model.lower():
+                continue
+            if table not in data.keys():
+                data[table] = []
+            data[table].append({"rule": {"id": rule.id, "fields": [{"id": field.id, "name": field.name} for field in rule_fields], "scopes": [{
+                               "id": _scope.id, "name": _scope.name} for _scope in scope], "isDatabaseConstraint": rule.isDatabaseConstraint}})
+
+    elif request.method == 'PUT' or request.method == 'POST':
+        rules = json.loads(request.body)['rules']
+        discipline = models.Discipline.objects.get(id=discipline_id)
+        for rule in rules:
+            fetched_scopes = models.Splocalecontaineritem.objects.filter(
+                id__in=[int(scope["id"]) for scope in rule["scopes"]]) if len(rule["scopes"]) > 0 else [None]
+            if rule["id"] is None:
+                fetched_rule = UniquenessRule.objects.create(
+                    isDatabaseConstraint=rule["isDatabaseConstraint"], discipline=discipline)
+            else:
+                fetched_rule = UniquenessRule.objects.get(id=rule["id"])
+                fetched_rule.discipline = discipline
+                fetched_rule.isDatabaseConstraint = rule["isDatabaseConstraint"]
+                fetched_rule.save()
+
+            fetched_rule.splocalecontaineritems.clear()
```

Comment:
Fixed in https://github.com/specify/specify7/pull/4215/commits/8c2eff7f85b62fd2c71ade8c10e933c9e5e5bfac#diff-3885e47982620743c313bf9b83cec4a1025f523e38d653ee72b38c52d1e6c123

Reply:
Hmm are you sure it is sufficient? Consider a rule (with id 1) with two fields - `field 1`, and `field 2`. User removes `field 2`. The current code will delete `field 1` (which is correct), but will not disable (`field 1` AND `field 2`) signal handler since rule (with id 1) has not been deleted, as this rule still exists.

It might be better if you simply delete a rule if any field got changed (added or deleted), and then make a new one?

Reply:
It may be because the comment is outdated, but we disconnect any existing signal handler for the rule before we set the new one. 
Is this not sufficient? It would ensure that everytime we call `make_uniqueness_rule` on a particular uniqueness rule that only the most recent changes to the rule are being tracked at a time. 
Also, `dispacth_uid` also ensures that only one instance of a signal with the same dispacth_uid is connected at a given time (see https://docs.djangoproject.com/en/3.2/topics/signals/#preventing-duplicate-signals)

Reply:
Yes, it should be enough! When I left the comment only the deleted rules were disconnected, thus the above review comment.